native/src/fairseq2n/data/dynamic_bucket_data_source.cc

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) Meta Platforms, Inc. and affiliates.
+// All rights reserved.
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include "fairseq2n/data/dynamic_bucket_data_source.h"
+
+#include <vector>
+
+namespace fairseq2n::detail {
+
+dynamic_bucket_data_source::dynamic_bucket_data_source(
+    std::unique_ptr<data_source> &&inner, float64 threshold, cost_fn &&fn, std::optional<std::size_t> maybe_nb_min, std::optional<std::size_t> maybe_nb_max, bool drop_remainder) noexcept
+  : inner_{std::move(inner)}, threshold_{threshold}, cost_fn_{std::move(fn)}, maybe_nb_min_{maybe_nb_min}, maybe_nb_max_{maybe_nb_max}, drop_remainder_{drop_remainder}
+{}
+
+std::optional<data>
+dynamic_bucket_data_source::next()
+{
+    data_list output{};
+
+    if (maybe_nb_min_)
+        output.reserve(*maybe_nb_min_);
+
+    float64 cost = 0;
+
+    auto bucket_ready = [&]() {
+        bool cost_threshold_met = (cost >= threshold_);
```

Comment:
Nit: In this line and the following two lines, you don't really need parenthesis in the expressions. Just:

```
bool cost_threshold_met = cost >= threshold_;
``

is enough.